mechanicalsoup/stateful_browser.py

```diff
@@ -207,20 +207,52 @@ def select_form(self, selector="form", nr=0):
         :return: The selected form as a soup object. It can also be
             retrieved later with the :attr:`form` attribute.
         """
+
+        def find_associated_elements(form_id):
+            """Function to find all associated elements to a form
+                (i.e. an element with a form attribute -> ``form=form_id``)
+            """
+
+            # Elements which can have a form owner
+            form_associated_elements = ("input", "button", "fieldset",
+                                        "object", "output", "select",
+                                        "textarea")
+
+            found_elements = []
+
+            for element in form_associated_elements:
+                for new_element in self.page.find_all(element,
+                                                      form=form_id):
+                    found_elements.append(new_element)
+
+            return found_elements
+
         if isinstance(selector, bs4.element.Tag):
             if selector.name != "form":
                 raise LinkNotFoundError
-            self.__state.form = Form(selector)
+            form = selector
         else:
             # nr is a 0-based index for consistency with mechanize
             found_forms = self.page.select(selector,
                                            limit=nr + 1)
+
+            if len(found_forms) > 0:
+                form = found_forms[-1]
+            else:
+                form = None
+
             if len(found_forms) != nr + 1:
                 if self.__debug:
```

Comment:
You want that error-checking code to be right after the call to `self.page.select(...)`, hence insert your code below this rather than above.